### PR TITLE
Infer conflicts based on regular expression inclusion

### PR DESCRIPTION
**Pull request by 4tXJ7f**
We have a conflict if we have `str.in.re(x, R1)` and `~str.in.re(x, R2)`
and `R2` includes `R1` because there is no possible value for `x` that
satisfies both memberships. This commit adds code to detect regular
expression inclusion for a small fragment of regular expressions: string
literals with single char (`re.allchar`) and multichar wildcards
(`re.*(re.allchar)`).

**NOTE**: Do not merge before I get approval from my employer.

### CODE REVIEW

File: src/theory/strings/regexp_operation.cpp [hunk 1]
```diff
@@ -1652,6 +1653,175 @@ std::string RegExpOpr::mkString( Node r ) {
...
+    }
+    newIdxs.clear();
+
+    if (n2.getKind() == STRING_TO_REGEXP || n2 == d_sigma)
+    {
+      Assert(n2 == d_sigma
+             || (n2[0].isConst() && n2[0].getConst<String>().size() == 1));
+      for (size_t idx : idxs)
+      {
+        if ((v1[idx] == d_sigma || v1[idx] == n2))
```
Comment (ajreynol) [inline src/theory/strings/regexp_operation.cpp]: Remove double parenthesis here.

File: src/theory/strings/regexp_solver.cpp [hunk 2]
```diff
@@ -271,6 +271,67 @@ void RegExpSolver::check(const std::map<Node, std::vector<Node> >& mems)
...
+    {
+      miNeg = mNeg;
+      miLit = mLit;
+    }
+    else
+    {
+      // Both regular expression memberships have the same polarity
+      if (mNeg == miNeg)
+      {
+        if (d_regexp_opr.regExpIncludes(miLit[1], mLit[1]))
```
[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/theory/strings/regexp_solver.cpp]: Great catch! I've added a set that keeps track of the removed items.

File: src/theory/strings/regexp_solver.cpp [hunk 3]
```diff
@@ -271,6 +271,67 @@ void RegExpSolver::check(const std::map<Node, std::vector<Node> >& mems)
...
+        {
+          d_parent.getExtTheory()->markReduced(miLit);
```
[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/theory/strings/regexp_solver.cpp]: Yup, done. I also added a regression that triggers those paths test.

File: src/theory/strings/regexp_solver.cpp [hunk 4]
```diff
@@ -271,6 +271,67 @@ void RegExpSolver::check(const std::map<Node, std::vector<Node> >& mems)
   }
 }
 
+bool RegExpSolver::checkEqcInclusion(const std::vector<Node>& mems)
+{
+  Node miLit;
+  bool miNeg = false;
+  for (const Node& m : mems)
```
[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/theory/strings/regexp_solver.cpp]: No idea what I was thinking here---not much apparently... Fixed.

File: src/theory/strings/regexp_solver.cpp [hunk 5]
```diff
@@ -271,6 +277,90 @@ void RegExpSolver::check(const std::map<Node, std::vector<Node> >& mems)
...
+
+          Node conc;
+          d_im.sendInference(vec_nodes, conc, "Intersect inclusion", true);
+          return false;
+        }
+      }
+    }
+  }
+
+  mems.erase(std::remove_if(
```
Comment (ajreynol) [inline src/theory/strings/regexp_solver.cpp]: nit: I'm not a huge fan of this coding style.  Would prefer a more basic way of writing this.

**Pull request merged**